packages/flutter/lib/src/material/pickers/calendar_date_picker.dart

```diff
@@ -0,0 +1,1050 @@
+      final String weekday = localizations.narrowWeekdays[i];
+      result.add(ExcludeSemantics(
+        child: Center(child: Text(weekday, style: headerStyle)),
+      ));
+      if (i == (localizations.firstDayOfWeekIndex - 1) % 7)
+        break;
+    }
+    return result;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final ThemeData theme = Theme.of(context);
+    final ColorScheme colorScheme = theme.colorScheme;
+    final TextStyle dayHeaderStyle = theme.textTheme.caption?.apply(
+      color: colorScheme.onSurface.withOpacity(0.60),
+    );
+    final MaterialLocalizations localizations = MaterialLocalizations.of(context);
+    final List<Widget> labels = _getDayHeaders(dayHeaderStyle, localizations);
+
+    return Padding(
+      padding: const EdgeInsets.symmetric(
+        horizontal: _monthPickerHorizontalPadding,
+      ),
+      child: GridView.custom(
+        shrinkWrap: true,
+        gridDelegate: _dayPickerGridDelegate,
+        childrenDelegate: SliverChildListDelegate(
+          labels,
+          addRepaintBoundaries: false,
+        ),
+      ),
+    );
+  }
+}
+
+/// A scrollable list of years to allow picking a year.
+class _YearPicker extends StatefulWidget {
+  /// Creates a year picker.
+  ///
+  /// The [currentDate, [firstDate], [lastDate], [selectedDate], and [onChanged]
+  /// arguments must be non-null. The [lastDate] must be after the [firstDate].
+  _YearPicker({
+    Key key,
+    @required this.currentDate,
+    @required this.firstDate,
+    @required this.lastDate,
+    @required this.initialDate,
+    @required this.selectedDate,
+    @required this.onChanged,
+  }) : assert(currentDate != null),
+       assert(firstDate != null),
+       assert(lastDate != null),
+       assert(initialDate != null),
+       assert(selectedDate != null),
+       assert(onChanged != null),
+       assert(!firstDate.isAfter(lastDate)),
+       super(key: key);
+
+  /// The current date.
+  ///
+  /// This date is subtly highlighted in the picker.
+  final DateTime currentDate;
+
+  /// The earliest date the user is permitted to pick.
+  final DateTime firstDate;
+
+  /// The latest date the user is permitted to pick.
+  final DateTime lastDate;
+
+  /// The initial date to center the year display around.
+  final DateTime initialDate;
+
+  /// The currently selected date.
+  ///
+  /// This date is highlighted in the picker.
+  final DateTime selectedDate;
+
+  /// Called when the user picks a year.
+  final ValueChanged<DateTime> onChanged;
+
+  @override
+  _YearPickerState createState() => _YearPickerState();
+}
+
+class _YearPickerState extends State<_YearPicker> {
+  ScrollController scrollController;
+
+  // The approximate number of years necessary to fill the available space.
+  static const int minYears = 18;
+
+  @override
+  void initState() {
+    super.initState();
+
+    // Set the scroll position to approximately center the initial year.
```

Comment:
Again, why approximate, and when does the approximation get resolved? Is there some kind of step that narrows down the value to the exact value?